### PR TITLE
Use Case: Mirror select command

### PR DESCRIPTION
**Pull request by hykilpikonna**
Please merge #69 first.

image

### CODE REVIEW

**Review by MstrPikachu — APPROVED** [review]

File: src/main/java/org/hydev/mcpm/client/arguments/parsers/MirrorParser.java [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+        return "Select a source (mirror) to download plugins from";
+    }
+
+    @Override
+    public void run(Namespace details, Consumer<String> log)
+    {
+        switch (details.getString("op"))
+        {
+            case "ping" -> controller.ping(details.getBoolean("refresh"), log);
+            case "sel" -> controller.sel(details.getString("host"), log);
```
Comment (1whatleytay) [inline src/main/java/org/hydev/mcpm/client/arguments/parsers/MirrorParser.java]: This is nice. I wish the method name was English though:
```suggestion
            case "sel" -> controller.select(details.getString("host"), log);
```

File: src/main/java/org/hydev/mcpm/client/commands/entries/MirrorController.java [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+{
+    /**
+     * Format ping with colors
+     *
+     * @param ping Ping in milliseconds
+     * @return Ping with colors
+     */
+    private static String formatPing(long ping)
+    {
+        var color = ping <= 100 ? "&a" : ping <= 200 ? "&e" : "&c";
```
Comment (1whatleytay) [inline src/main/java/org/hydev/mcpm/client/commands/entries/MirrorController.java:27]: Embedded ternary is sane but hard to read. If this is going to be its own method I would expand this and `formatSpeed` out into at least one if statement.

File: src/main/java/org/hydev/mcpm/client/database/LocalPluginTracker.java [hunk 3]
```diff
@@ -25,7 +26,7 @@
  */
 public class LocalPluginTracker implements PluginTracker {
     // CSV file storing the list of manually installed plugins
-    private String mainLockFile = "plugins/mcpm.lock.csv";
+    private String mainLockFile = "plugins/mcpm.lock.json";
```
Comment (1whatleytay) [inline src/main/java/org/hydev/mcpm/client/database/LocalPluginTracker.java]: How did this sneak its way into another PR? LOL
LMK when you revert this.

File: src/main/java/org/hydev/mcpm/client/database/mirrors/MirrorSelector.java [hunk 4]
```diff
@@ -47,4 +50,18 @@ public void updateMirrors() throws IOException
...
+    {
+        if (LOCAL_SEL.exists()) return YML.readValue(Files.readString(LOCAL_SEL.toPath()), new TypeReference<>() {});
+        return listAvailableMirrors().get(0);
+    }
+
+    @Override
+    public void setSelectedMirror(Mirror mirror) throws IOException
+    {
+        LOCAL_SEL.getParentFile().mkdirs();
+        Files.writeString(LOCAL_SEL.toPath(), YML.writeValueAsString(mirror));
```
Comment (1whatleytay) [inline src/main/java/org/hydev/mcpm/client/database/mirrors/MirrorSelector.java:65]: What level is MirrorSelector.java in CA? Is it database and frameworks? Makes sense.

**Review by 1whatleytay — CHANGES_REQUESTED** [review]
Requesting changes because of the LocalPluginTracker (sorry). Everything else looks good.

**Comment by hykilpikonna** [comment]
> Requesting changes because of the LocalPluginTracker (sorry). Everything else looks good.

Okay fixed

File: src/main/java/org/hydev/mcpm/client/database/LocalPluginTracker.java (hunk 3 above)
Comment (hykilpikonna) [reply src/main/java/org/hydev/mcpm/client/database/LocalPluginTracker.java]: fixed. I just have to rebase from main

**Review by 1whatleytay — APPROVED** [review]

File: src/main/java/org/hydev/mcpm/client/database/mirrors/MirrorSelector.java (hunk 4 above)
Comment (hykilpikonna) [reply src/main/java/org/hydev/mcpm/client/database/mirrors/MirrorSelector.java:65]: Yea I think it should be

File: src/main/java/org/hydev/mcpm/client/arguments/parsers/MirrorParser.java (hunk 1 above)
Comment (hykilpikonna) [reply src/main/java/org/hydev/mcpm/client/arguments/parsers/MirrorParser.java]: ok fixed

**Pull request merged**